src/libcore/iter/traits/iterator.rs

```diff
@@ -234,11 +234,16 @@ pub trait Iterator {
     /// assert_eq!(a.iter().count(), 5);
     /// ```
     #[inline]
-    #[rustc_inherit_overflow_checks]
     #[stable(feature = "rust1", since = "1.0.0")]
     fn count(self) -> usize where Self: Sized {
         // Might overflow.
-        self.fold(0, |cnt, _| cnt + 1)
+        #[inline]
+        #[rustc_inherit_overflow_checks]
+        fn add1<T>(count: usize, _: T) -> usize {
+            count + 1
```

Comment:
Might be worth double-checking there's a test for overflow checks here?  I'm never confident in what `rustc_inherit_overflow_checks` does.  (And I hear that one can use `Add::add(count, 1)` instead of using the attribute.)